src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java

```diff
@@ -29,62 +31,61 @@
 import software.amazon.awssdk.iot.iotshadow.model.UpdateShadowRequest;
 import software.amazon.awssdk.services.greengrassv2data.model.ConnectivityInfo;
-import software.amazon.awssdk.services.greengrassv2data.model.InternalServerException;
-import software.amazon.awssdk.services.greengrassv2data.model.ThrottlingException;
 
 import java.time.Duration;
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.Callable;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingDeque;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 import javax.inject.Inject;
 
 @SuppressWarnings("PMD.ImmutableField")
 public class CISShadowMonitor implements Consumer<NetworkStateProvider.ConnectionState> {
     private static final Logger LOGGER = LogManager.getLogger(CISShadowMonitor.class);
+    private static final String KV_TASK = "task";
     private static final String CIS_SHADOW_SUFFIX = "-gci";
-    private static final String VERSION = "version";
-
-    private static final RetryUtils.RetryConfig GET_CONNECTIVITY_RETRY_CONFIG =
-            RetryUtils.RetryConfig.builder().initialRetryInterval(Duration.ofMinutes(1L))
-                    .maxRetryInterval(Duration.ofMinutes(30L)).maxAttempt(Integer.MAX_VALUE)
-                    .retryableExceptions(Arrays.asList(ThrottlingException.class, InternalServerException.class))
-                    .build();
     private static final RetryUtils.RetryConfig ALWAYS_RETRY_CONFIG =
             RetryUtils.RetryConfig.builder()
                     .initialRetryInterval(Duration.ofSeconds(1L))
                     .maxRetryInterval(Duration.ofSeconds(30L))
                     .maxAttempt(Integer.MAX_VALUE)
                     .retryableExceptions(Collections.singletonList(Exception.class))
                     .build();
+    private final CISShadowTaskExecutor taskExecutor = new CISShadowTaskExecutor();
+    private final CISShadowTaskQueue taskQueue = new CISShadowTaskQueue();
 
-    private final Consumer<ShadowDeltaUpdatedEvent> onShadowDeltaUpdated = this::processCISShadow;
+    private final Consumer<ShadowDeltaUpdatedEvent> onShadowDeltaUpdated = resp ->
+            taskQueue.offer(new ProcessCISShadowTask(
+                    resp.version, Coerce.toString(resp.state.get("version")), resp.state));
```

Comment:
resp.state may be null. resp.state.get("version") definitely can be null.

Reply:
ah right, was thinking shadow state would always be correct, will fix

Reply:
fixed